pkg/app/piped/controller/scheduler.go

```diff
@@ -345,10 +345,53 @@ func (s *scheduler) Run(ctx context.Context) error {
 			))
 			defer span.End()
 
+			s.notifier.Notify(model.NotificationEvent{
+				Type: model.NotificationEventType_EVENT_STAGE_STARTED,
+				Metadata: &model.NotificationEventStageStarted{
+					Deployment: s.deployment,
+					Stage:      ps,
+				},
+			})
+
 			result = s.executeStage(sig, *ps, func(in executor.Input) (executor.Executor, bool) {
 				return s.executorRegistry.Executor(model.Stage(ps.Name), in)
 			})
 
+			switch result {
+			case model.StageStatus_STAGE_SUCCESS, model.StageStatus_STAGE_EXITED: // Exit stage is treated as success.
+				s.notifier.Notify(model.NotificationEvent{
+					Type: model.NotificationEventType_EVENT_STAGE_SUCCEEDED,
+					Metadata: &model.NotificationEventStageSucceeded{
+						Deployment: s.deployment,
+						Stage:      ps,
+					},
+				})
+			case model.StageStatus_STAGE_FAILURE:
+				s.notifier.Notify(model.NotificationEvent{
+					Type: model.NotificationEventType_EVENT_STAGE_FAILED,
+					Metadata: &model.NotificationEventStageFailed{
+						Deployment: s.deployment,
+						Stage:      ps,
+					},
+				})
+			case model.StageStatus_STAGE_CANCELLED:
+				s.notifier.Notify(model.NotificationEvent{
+					Type: model.NotificationEventType_EVENT_STAGE_CANCELLED,
+					Metadata: &model.NotificationEventStageCancelled{
+						Deployment: s.deployment,
+						Stage:      ps,
+					},
+				})
+			case model.StageStatus_STAGE_SKIPPED:
+				s.notifier.Notify(model.NotificationEvent{
+					Type: model.NotificationEventType_EVENT_STAGE_SKIPPED,
+					Metadata: &model.NotificationEventStageSkipped{
+						Deployment: s.deployment,
+						Stage:      ps,
+					},
+				})
+			}
+
```

Comment:
[IMO] How about extracting into a method? 
I just want to simplify `scheduler.Run()` for readability and maintenanceability.

For example: 

```suggestion
			s.notifyStageEvent(model.StageStatus_STAGE_STARTED, ps)

			result = s.executeStage(sig, *ps, func(in executor.Input) (executor.Executor, bool) {
				return s.executorRegistry.Executor(model.Stage(ps.Name), in)
			})

                         s.notifyStageEvent(result, ps)
```

Reply:
Thank you!
StageStart and StageEnd are really good.